providers/fips/selftest.c

```diff
@@ -10,6 +10,17 @@
 #include <string.h>
 #include <openssl/evp.h>
 #include <openssl/params.h>
+#include <openssl/crypto.h>
+#include "e_os.h"
+/*
+ * We're cheating here. Normally we don't allow RUN_ONCE usage inside the FIPS
+ * module because all such initialisation should be associated with an
+ * individual OPENSSL_CTX. That doesn't work with the self test though because
+ * it should be run once regardless of the number of OPENSSL_CTXs we have.
+ */
+#undef FIPS_MODE
+#include <internal/thread_once.h>
+#define FIPS_MODE
 #include "selftest.h"
```

Comment:
I don't like this sort of approach of switching off the context to bypass a check - we should be defining an additional flag and allowing the check to pass rather than hiding our context here ...

Reply:
I've changed this so that it sets a new define prior to the inclusion

Reply:
What happens if we just allow `RUN_ONCE` unconditionally in the FIPS module?

Reply:
> What happens if we just allow RUN_ONCE unconditionally in the FIPS module?

In the vast majority of cases (all of them so far except this one) its not what you want to do. It's a bit of pandora's box wrt freeing up the memory afterwards. It turns out to be hard to do since there is no atexit() in the FIPS provider, and there is no portable way of doing a ".fini". (you may recall the long conversations we had about freeing resources (ref counting etc) when we first started implementing this stuff in the FIPS module). We only get away with it here because we take the explicit decision *not* to free up the lock if we're on a platform where we don't know how to do it.

We just avoid all of those issues if we disallow RUN_ONCE - and it turns out to work quite well without it. Instead resources are not allocated on a per-DSO basis, but on a per OPENSSL_CTX basis. We get a new OPENSSL_CTX every time OSSL_init_provider gets called. And then we free up the OPENSSL_CTX and all the associated resources when the teardown function gets called. All the code that historically was being done via RUN_ONCE can be better done as part of the OPENSSL_CTX init.